### PR TITLE
Batch transformations. #429

### PR DESCRIPTION
**Pull request by ake123**
Batch transformations. #429

### CODE REVIEW

**Review by TuomasBorman — CHANGES_REQUESTED** [review]
I think this is too complex, instead you could consider own method for SingleCellExperiment. Something like this

SCE_method <- function(x, altexp = altExpNames(tse), ...){
    # Check that specified altexps are found from x
    # Check that assay.type are found from specified altexps
    lapply(altExps(x)[altexp], .check_assay_present)
    # Apply transformation to x
    x <- .perform_transformation(x)
    # If altexp is not NULL, apply transformation to altexps
    altExps(x)[altexp] <- bplapply(altExps(x)[altexp], ...)
}

SE_method <- function(){
    # Input checks
    # Apply transformation
    x <- .perform_transformation(x)
}

**Comment by antagomir** [comment]
I agree that direct support for SCE is probably the way to go when altExps are supported. 

This in the above example mixes concepts: `altexp = reducedDimanames(tse)`? Should it be `altexp = altExpNames(tse)`?

**Comment by TuomasBorman** [comment]
> I agree that direct support for SCE is probably the way to go when altExps are supported.
> 
> This in the above example mixes concepts: `altexp = reducedDimanames(tse)`? Should it be `altexp = altExpNames(tse)`?

Yep, I fixed the comment

**Comment by antagomir** [comment]
Status of this one?

**Comment by ake123** [comment]
rewriting and debugging it as suggested

File: R/transformCounts.R [hunk 1]
```diff
@@ -123,86 +134,121 @@ setGeneric("transformAssay", signature = c("x"),
...
+                        assays = SimpleList(counts = matrix(NA, nrow = nrow(x), ncol = ncol(x))),
+                        rowData = rowData(x),
+                        colData = colData(x)
+                    )
+                    # Add this new altExp to the SingleCellExperiment
+                    altExp(x, alt_name) <- new_altExp
+                }
+            }
+        }
+
```
Comment (TuomasBorman) [inline R/transformCounts.R]: This is much more complex than should be

File: R/transformCounts.R [hunk 2]
```diff
@@ -240,17 +286,59 @@ setMethod("transformAssay", signature = c(x = "SummarizedExperiment"),
...
+#     if(identical(rownames(transformed_table), colnames(mat)) &&
+#        identical(colnames(transformed_table), rownames(mat)) &&
+#        ncol(transformed_table) != ncol(mat) &&
+#        nrow(transformed_table != nrow(mat))){
+#         transformed_table <- t(transformed_table)
+#     }
+#     return(transformed_table)
+# }
+
 .apply_transformation_from_vegan <- function(mat, method, MARGIN, reference = ref_vals,
```
Comment (TuomasBorman) [inline R/transformCounts.R]: Why the function is duplicated?

**Comment by TuomasBorman** [comment]
Can you revert the already made modifications? Something like this creates a good start point

```
#' @rdname transformAssay
#' @export
setGeneric("transformAssay", signature = c("x"),
    function(x,  ...)
    standardGeneric("transformAssay"))

#' @rdname transformAssay
#' @export
setMethod("transformAssay", signature = c(x = "SummarizedExperiment"),
    function(x,
        assay.type = "counts", assay_name = NULL,
        method = c("alr", "chi.square", "clr", "css", "css.fast", "frequency", 
                  "hellinger", "log", "log10", "log2", "max", "normalize", 
                  "pa", "range", "rank", "rclr", "relabundance", "rrank",
                  "standardize", "total", "z"),
        MARGIN = "samples",
        name = method,
        pseudocount = FALSE,
        ...){
        #
        x <- .transform_assay(
            x = x, method = method, name = name, assay.type = assay.type,
            MARGIN = MARGIN, pseudocount =  pseudocount, ...)
        return(x)
    }
)

.transform_assay <- function(
        x, assay.type = "counts", assay_name = NULL,
        method = c(
            "alr", "chi.square", "clr", "css", "css.fast", "frequency", 
            "hellinger", "log", "log10", "log2", "max", "normalize", 
            "pa", "range", "rank", "rclr", "relabundance", "rrank",
            "standardize", "total", "z"),
         MARGIN = "samples",
         name = method,
         pseudocount = FALSE,
        ...){
    # Input check
    if (!is.null(assay_name)) {
        .Deprecated(old="assay_name", new="assay.type", "Now assay_name is 
                deprecated. Use assay.type instead.")
        assay.type <- assay_name
    }
    
    # Check assay.type
    .check_assay_present(assay.type, x)
    
    # Check name
    if(!.is_non_empty_string(name) ||
       name == assay.type){
        stop("'name' must be a non-empty single character value and be ",
             "different from `assay.type`.",
             call. = FALSE)
    }
    # Check method
    # If method is not single string, user has not specified transform method,
    # or has given e.g. a vector
    if(!.is_non_empty_string(method)){
        stop("'method' must be a non-empty single character value.",
             call. = FALSE)
    }
    method <- match.arg(method, several.ok = FALSE)
    # Check that MARGIN is 1 or 2
    MARGIN <- .check_MARGIN(MARGIN)
    # Check pseudocount
    if( !.is_a_bool(pseudocount) && !(is.numeric(pseudocount) && 
                                      length(pseudocount) == 1 && pseudocount >= 0) ){
        stop("'pseudocount' must be TRUE, FALSE or a number equal to or 
                 greater than 0.",
             call. = FALSE)
    }
    # Input check end
    #
    # Get the method and abundance table
    method <- match.arg(method)
    assay <- assay(x, assay.type)
    
    # Apply pseudocount, if it is not 0
    assay <- .apply_pseudocount(assay, pseudocount, ...)
    # Store pseudocount value and set attr equal to NULL
    pseudocount <- attr(assay, "pseudocount")
    attr(assay, "pseudocount") <- NULL
    
    # Calls help function that does the transformation
    # Help function is different for mia and vegan transformations
    if( method %in% c("log10", "log2", "css", "css.fast") ){
        transformed_table <- .apply_transformation(
            assay, method, MARGIN, ...)
    } else {
        transformed_table <- .apply_transformation_from_vegan(
            assay, method, MARGIN, ...)
    }
    
    # Add pseudocount info to transformed table
    attr(transformed_table, "parameters")$pseudocount <- pseudocount
    
    # Assign transformed table to assays
    assay(x, name, withDimnames=FALSE) <- transformed_table
    x
}

#' @rdname transformAssay
#' @export
setMethod("transformAssay", signature = c(x = "SingleCellExperiment"),
    function(x, altexp = altExpNames(x), ...){
        # Check altexp
        if( !(is.null(altexp) || (!is.null(altexp) && all(altexp %in% altExpNames(x)))) ){
            stop(".", call. = FALSE)
        }
        browser()
        # Transform the main object
        x <- .transform_assay(x, ...)
        altExps(x)[altexp] <- lapply(altExps(x)[altexp], function(y){
            .transform_assay(y, ...)
        })
        return(x)
    }
)
```

File: R/transformCounts.R [hunk 3]
```diff
@@ -109,97 +114,132 @@
 #' assay(tse, "rank_average", withDimnames = FALSE) <- colRanks(
 #'     assay(tse, "counts"), ties.method = "average", preserveShape = TRUE)  
 #' 
+#' #using altexp parameter
+#' data("GlobalPatterns", package="mia")
+#' tse <- GlobalPatterns
+#' altExp(tse,"species") <- transformAssay(tse, method = "relabundance",rank = "species", na.rm = FALSE)
+#' altExp(tse,"species")
```
Comment (TuomasBorman) [inline R/transformCounts.R]: Add space after #

Comment (TuomasBorman) [reply R/transformCounts.R]: I made the changes, see those

File: R/transformCounts.R [hunk 4]
```diff
@@ -109,97 +114,132 @@
 #' assay(tse, "rank_average", withDimnames = FALSE) <- colRanks(
 #'     assay(tse, "counts"), ties.method = "average", preserveShape = TRUE)  
 #' 
+#' #using altexp parameter
+#' data("GlobalPatterns", package="mia")
+#' tse <- GlobalPatterns
+#' altExp(tse,"species") <- transformAssay(tse, method = "relabundance",rank = "species", na.rm = FALSE)
```
Comment (TuomasBorman) [inline R/transformCounts.R]: There is no rank parameter in transformAssay()

File: R/transformCounts.R [hunk 5]
```diff
@@ -241,16 +281,15 @@ setMethod("transformAssay", signature = c(x = "SummarizedExperiment"),
...
-        stop("'reference' must be a single value specifying the ",
-             "values of the reference sample.",
-             call. = FALSE)
+                                            ref_vals = NA, ...) {
+    # Ensure that the matrix has proper dimnames
+    if (is.null(rownames(mat))) {
+        rownames(mat) <- paste0("feature", seq_len(nrow(mat)))
     }
-    # Input check end
-
```
Comment (TuomasBorman) [inline R/transformCounts.R]: Why this was removed?

**Review by TuomasBorman — CHANGES_REQUESTED** [review]
Can you still add unit tests, that altexp works as expected?

File: R/transformCounts.R (hunk 4 above)
Comment (TuomasBorman) [reply R/transformCounts.R]: I made the changes

**Review by TuomasBorman — APPROVED** [review]

**Pull request merged**